### ISSUE TITLE
KeepLinkedVertices

### ISSUE THREAD

**Issue by amirebrahimi** [Whinarn/MeshDecimator#2]
I'm encountering quite a few models with this model pack that don't reduce unless I set `KeepLinkedVertices = false`. I don't see this in https://github.com/sp4cerat/Fast-Quadric-Mesh-Simplification. Can you share a bit more about what you're doing here? I assume that you are preserving duplicate vertices, so the model doesn't have holes? Did you think about welding them perhaps instead?

**Comment by Whinarn** [comment]
You are correct that it was an attempt to prevent holes. The idea was that vertices at the same position but part of different triangles would be kept, if I remember this correctly. Some of those vertices might have differerent UVs etc which would make the decimated result look quite horrible in cases where I tried to merge them, so I ended up with the approach that I called linking vertices (perhaps a bad name for it?).

This was not a part of https://github.com/sp4cerat/Fast-Quadric-Mesh-Simplification as you say.

It's a long time ago since I worked on this, so I will read through the code in a few hours and get back to you on this. I am very sure that this can be improved.

**Comment by Whinarn** [comment]
Upon further inspection and testing I found the KeepLinkedVertices rather ineffective. I couldn't find any remarkable differences from having KeepBorder turned on (on different meshes I found). If I remember correctly it was an experiment for an alternative to KeepBorder but with the possibility to reduce the vertex count further. But it is very bad at that and doesn't solve the problem either. I suppose I just forgot it there after.

I will look into welding vertices, because that should allow the decimation to be a lot more powerful.

**Comment by amirebrahimi** [comment]
My suggestion of welding vertices is not correct since if we welded vertices, then we would lose unique UVs. Doing a little research online it seems someone else came across this, too:
https://github.com/libigl/libigl/issues/378

Instead of skipping linked vertices it might be a matter of keeping track of coincident vertices in space and factoring that into the error metric / edge collapse cost somehow if they are not exactly the same in all regard (UVs, normals, etc.).

It's too bad that https://github.com/sp4cerat/Fast-Quadric-Mesh-Simplification didn't attempt to solve this issue as it is essential for any modern-day decimation.

**Comment by Whinarn** [comment]
Sorry for not responding in a few days. I have been looking into different approaches and found this [article ](https://cragl.cs.gmu.edu/seamless/) linking to this [repository](https://github.com/songrun/SeamAwareDecimater).

I have been experimenting a little bit, but without much success so far.

I will try to dig deeper into it more when I have some more time to spare.

**Comment by amirebrahimi** [comment]
No need to apologize, you're not on the hook for anything ;)

Good find though and as a first stab, I may try to generate some bindings in C# for the repo using [CppSharp](https://github.com/mono/CppSharp) or [Swig](https://github.com/swig/swig) just to evaluate. If it works well long-term, then it might make more sense to create a C# implementation.

What were your thoughts? Have you been trying to modify your existing implementation to accomodate the changes introduced in the paper?

**Comment by Whinarn** [comment]
So far I have only spent a few hours in the weekend trying different ideas of my own, not related to the article I found.

I was interested in porting https://github.com/songrun/SeamAwareDecimater to C# however. So I would be interested to hear about your results if you generate bindings for it.

**Comment by amirebrahimi** [comment]
Catching up on my backlog of AutoLOD related issues. I had saved this to come back to. What's your take on the SeamAwareDecimater now that you've added Smart Linking?

**Comment by Whinarn** [comment]
I apparently forgot to mention it, but I did end up trying out the SeamAwareDecimator and found it really slow. There's of course still a chance to implement support for that algorithm. But one of the goals that I had with this project was to provide quick and decent quality mesh decimation.

The SeamAwareDecimator inspired me when creating the smart linking feature. So therefore I do not feel any urgency to add support for the SeamAwareDecimator. It's worth to add here that while I made the 
FastQuadricMeshSimplification algorithm seam-aware with this feature, it mainly aims to prevent holes. The UV distribution could be greatly improved still, which was something that the SeamAwareDecimator was designed to take care of.

As for the KeepLinkedVertices it will be depricated in my next release.
